brave/src/main/java/brave/propagation/B3SingleFormat.java

```diff
@@ -140,101 +140,109 @@ public static TraceContextOrSamplingFlags parseB3SingleFormat(CharSequence b3) {
+    long traceIdHigh, traceId;
+    int traceIdLength = indexOfFirstHyphen - beginIndex;
+    if (traceIdLength == 32) {
+      traceIdHigh = lenientLowerHexToUnsignedLong(value, pos, pos + 16);
+      pos += 16; // upper 64 bits of the trace ID
+    } else if (traceIdLength == 16) {
+      traceIdHigh = 0L;
+    } else {
+      Platform.get().log("Invalid input: expected a 16 or 32 lower hex trace ID", null);
       return null;
     }
+
+    traceId = tryParseId("trace", value, pos, endIndex);
+    if (traceId == 0) return null;
+    pos += 17; // lower 64 bits of the trace ID and the hyphen
+
+    long spanId = tryParseId("span", value, pos, endIndex);
+    if (spanId == 0) return null;
     pos += 16; // spanid
 
     int flags = 0;
     long parentId = 0L;
     if (endIndex > pos) {
-      if (isLowerHex(b3.charAt(pos))) {
-        Platform.get().log("Invalid input: span ID is too long", null);
+      pos++; // consume the delimiter
+
+      if (endIndex == pos) { // traceid-spanid-
+        Platform.get().log("Truncated after reading span ID", null);
         return null;
       }
 
-      // If we are at this point, we have more than just traceId-spanId.
-      // If the sampling field is present, we'll have a delimiter 2 characters from now. Ex "-1"
-      // If it is absent, but a parent ID is (which is strange), we'll have at least 17 characters.
-      // Therefore, if we have less than two characters, the input is truncated.
-      if (endIndex == pos + 1) {
-        Platform.get().log("Invalid input: truncated", null);
+      char nextChar = value.charAt(pos);
+      if (nextChar == '-') { // traceid-spanid--...
+        Platform.get().log("Invalid input: empty sampled field", null);
         return null;
       }
-      if (!checkHyphen(b3, pos++)) return null;
 
-      // If our position is at the end of the string, or another delimiter is one character past our
-      // position, try to read sampled status.
-      boolean afterSampledField = notHexFollowsPos(b3, pos, endIndex);
-      if (endIndex == pos + 1 || afterSampledField) {
-        flags = parseFlags(b3, pos);
+      // At this point, we are either at the sampling field, or first character of the parentId
+      int nextPos = pos + 1;
+      boolean hasSampledAndParent = hyphenCount == 3;
+      if (nextPos == endIndex || hasSampledAndParent) {
+        flags = parseSampledFlags(nextChar);
+        pos++; // consume the sampled field
         if (flags == 0) return null;
-        pos++; // consume the sampled status
-        if (afterSampledField && !checkHyphen(b3, pos++)) return null; // consume the delimiter
+        if (hasSampledAndParent && value.charAt(pos++) != '-') {
+          Platform.get().log("Invalid input: sampled is too long", null);
+          return null;
+        }
       }
 
-      if (endIndex > pos || afterSampledField) {
+      if (endIndex > nextPos) {
```

Comment:
Think this is supposed to be `pos` - if not then minutes of staring still hasn't let my brain fully grok `nextPos`

Reply:
it isn't pos.. anyway I'll try to rename things

Reply:
tried a different approach. PTAL